src/function.ts

```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+              } else {
+                tokens = [...tokens, reversed.toString()];
+              }
+            }
+          }
+          return true;
+        },
+      }
+    );
+
+    const tokenContext = tokens.map((t) => {
+      const id = new RegExp("\\${Token\\[TOKEN\\.([0-9]*)\\]}", "g").exec(
+        t
+      )?.[1];
+      return {
+        id,
+        token: t,
+        // env variables must start with a alpha character
+        env: `env__functionless${id}`,
+      };
+    });
+
+    const resultText = tokenContext.reduce(
+      // TODO: support templated strings
+      (r, t) => r.split(`"${t.token}"`).join(`process.env.${t.env}`),
+      result.text
+    );
+
+    // console.log(resultText);
+
+    const asset = aws_lambda.Code.fromAsset("", {
+      assetHashType: AssetHashType.OUTPUT,
+      bundling: {
+        image: DockerImage.fromRegistry("empty"),
+        // This forces the bundle directory and cache key to be unique. It does nothing else.
+        user: scope.node.addr,
+        local: {
+          tryBundle(outdir: string) {
+            fs.writeFileSync(path.resolve(outdir, "index.js"), resultText);
+            return true;
+          },
+        },
+      },
+    });
+
+    if (!(scope instanceof aws_lambda.Function)) {
+      throw new Error(
+        "CallbackLambdaCode can only be used on aws_lambda.Function"
+      );
+    }
+
+    tokenContext.forEach((t) => scope.addEnvironment(t.env, t.token));
+
+    const funcResource = scope.node.findChild(
+      "Resource"
+    ) as aws_lambda.CfnFunction;
+
+    const codeConfig = asset.bind(scope);
+
+    funcResource.code = {
+      s3Bucket: codeConfig.s3Location?.bucketName,
+      s3Key: codeConfig.s3Location?.objectKey,
+      s3ObjectVersion: codeConfig.s3Location?.objectVersion,
+      zipFile: codeConfig.inlineCode,
+      imageUri: codeConfig.image?.imageUri,
+    };
+
+    asset.bindToResource(funcResource);
+  }
+}
+
+export interface NativeIntegration<F extends AnyFunction> {
+  bootstrap: (context: Function<any, any>) => void;
+  /**
+   * @param args The arguments passed to the integration function by the user.
+   * @param preWarmContext contains singleton instances of client and other objects initialized outside of the native
+   *                       function handler.
+   */
+  call: (
+    args: Parameters<F>,
+    preWarmContext: NativePreWarmContext
+  ) => Promise<ReturnType<F>>;
+  /**
+   * Method called outside of the handler to initialize things like the PreWarmContext
+   */
+  preWarm?: (preWarmContext: NativePreWarmContext) => void;
+}
+
+/* istanbul ignore next */
+export class NativePreWarmContext {
+  private readonly cache: Record<string, any>;
+
+  constructor() {
+    this.cache = {};
+  }
+
+  public lambda(): Lambda {
+    return this.getOrInit("LAMBDA", () => new Lambda());
+  }
+
+  public eventBridge(): EventBridge {
+    return this.getOrInit("EVENT_BRIDGE", () => new EventBridge());
+  }
```

Comment:
I like how this class caches things, but is it a good idea to embed the individual clients in here? Won't that harm tree-shaking?

Reply:
Good question, answered above that the impact is low due to the use of `require("aws_sdk")`, but I would also like to reduce the size of the outputted code in any way...

The problem I had was... any integration may need any client, and I wouldn't want a client to add it's own flavor behind a shared key like `this.getOrInit("LAMBDA", () => new SpecialLambdaClient())`.

I guess we could make constants that encode the shared logic.

```ts
prewarmClient.get(PrewarmClient.LAMBDA);

// I am not sure how well an enum would serialize either
enum PrewarmClient {
   LAMBDA: () => new Lambda();
}
```

Will play with it.

Reply:
Got it

```ts
      preWarm: (preWarmContext: NativePreWarmContext) => {
        preWarmContext.getOrInit(PrewarmClients.LAMBDA);
      },
```

Serialized, only the required methods: 

```ts
var __exports_PrewarmClients = {};
var __exports_PrewarmClients_LAMBDA = {key: "LAMBDA", init: __f2};
__exports_PrewarmClients.LAMBDA = __exports_PrewarmClients_LAMBDA;
__exports.PrewarmClients = __exports_PrewarmClients;
```